validator/src/server.rs

```diff
@@ -0,0 +1,463 @@
+//! Validator gRPC high-level server implementation
+
+use std::collections::HashMap;
+
+use tendermint::{
+    block, chain, consensus,
+    proposal::{SignProposalRequest, SignedProposalResponse},
+    public_key::{PubKeyRequest, PubKeyResponse},
+    vote::{SignVoteRequest, SignedVoteResponse},
+    PublicKey, Signature,
+};
+use tendermint_proto::privval::{
+    priv_validator_api_server::{PrivValidatorApi, PrivValidatorApiServer},
+    PubKeyRequest as RawPubKeyRequest, PubKeyResponse as RawPubKeyResponse, RemoteSignerError,
+    SignProposalRequest as RawSignProposalRequest, SignVoteRequest as RawSignVoteRequest,
+    SignedProposalResponse as RawSignedProposalResponse,
+    SignedVoteResponse as RawSignedVoteResponse,
+};
+use tokio::{net::UnixListener, sync::Mutex, time::Instant};
+use tokio_stream::wrappers::UnixListenerStream;
+use tonic::{transport::Server, Request, Response, Status};
+use tracing::{debug, error, info};
+
+use crate::{
+    config::{BasicServerConfig, GrpcSocket},
+    signer::{display_validator_info, SignerProvider},
+    state::ValidatorStateProvider,
+};
+
+/// Validator gRPC high-level server implementation
+#[derive(Debug)]
+pub struct KMSServer<S: SignerProvider, VS: ValidatorStateProvider> {
+    /// Signer and state providers for individual chains.
+    /// Given the traits require mutable access to the state provider
+    /// across awaits, we use tokio's mutex here.
+    providers: Mutex<HashMap<chain::Id, (S, VS)>>,
+    /// A cache of public keys loaded during the initialization.
+    pubkeys: HashMap<chain::Id, PublicKey>,
+    /// Optional setting for the maximum height,
+    /// after which the server will stop signing for a particular network.
+    max_heights: HashMap<chain::Id, block::Height>,
```

Comment:
These seem like cases that don't benefit from an unordered map.

Perhaps use a `BTreeMap` instead, which provides deterministic ordering?